mechanicalsoup/browser.py

```diff
@@ -149,53 +150,54 @@ def _request(self, form, url=None, **kwargs):
-
-        for textarea in form.select("textarea"):
-            name = textarea.get("name")
-            if not name:
-                continue
-            data[name] = textarea.text
-
-        for select in form.select("select"):
-            name = select.get("name")
-            if not name:
-                continue
-            multiple = "multiple" in select.attrs
-            values = []
-            for i, option in enumerate(select.select("option")):
-                if (i == 0 and not multiple) or "selected" in option.attrs:
-                    values.append(option.get("value", ""))
-            if multiple:
-                data[name] = values
-            elif values:
-                data[name] = values[-1]
+        # Use a list of 2-tuples to better reflect the behavior of browser QSL.
+        # Requests also retains order when encoding form data in 2-tuple lists.
+        data = [(k, v) for k, v in data.items()]
+
+        # Process form tags in the order that they appear on the page,
+        # skipping those tags that do not have a name-attribute.
+        selector = ",".join("{}[name]".format(i) for i in
+                            ("input", "button", "textarea", "select"))
+        for tag in form.select(selector):
+            name = tag.get("name")  # name-attribute of tag
+
+            if tag.name in ("input", "button"):
+                if tag.get("type") in ("radio", "checkbox"):
+                    if "checked" not in tag.attrs:
+                        continue
+                    value = tag.get("value", "on")
+                else:
+                    # browsers use empty string for inputs with missing values
+                    value = tag.get("value", "")
+
+                if tag.get("type") == "file":
+                    # read http://www.cs.tut.fi/~jkorpela/forms/file.html
+                    # in browsers, file upload only happens if the form
+                    # (or submit button) enctype attribute is set to
+                    # "multipart/form-data". We don't care, simplify.
+                    if not value:
+                        continue
+                    if isinstance(value, string_types):
+                        value = open(value, "rb")
+                    files[name] = value
+                else:
+                    data.append((name, value))
+
+            elif tag.name == "textarea":
+                data.append((name, tag.text))
+
+            elif tag.name == "select":
+                multiple = "multiple" in tag.attrs
+                values = []
+                for i, option in enumerate(tag.select("option")):
+                    if (i == 0 and not multiple) or "selected" in option.attrs:
+                        values.append(option.get("value", ""))
+                if multiple:
+                    for value in values:
+                        data.append((name, value))
+                elif values:
+                    # Selects the first option if none are selected
```

Comment:
I think the comment is misplaced. It should be near the `if (i==0 and not multiple)` above. And actually, I think the code is a bit trickier than it needs to be. I'd just do something like

```
                options = tag.select("option")
                for i, option in enumerate(options):
                    if "selected" in option.attrs:
                        values.append(option.get("value", ""))
```

and then

```
                elif values:
                    data.append((name, values[-1]))
                elif options:
                    # Select the first option if none are selected
                    data.append((name, options[0].get("value", "")))
```

(Untested, but should be more self-explanatory and not much longer)

Reply:
I was quite confused by this code originally (not least because of the complete lack of comments), so I am very happy to rework it. I took your idea here and made a few modifications. How does this look to you?
```python
            elif tag.name == "select":
                options = tag.select("option")
                selected_values = [i.get("value", "") for i in options
                                   if "selected" in i.attrs]
                if "multiple" in tag.attrs:
                    for value in selected_values:
                        data.append((name, value))
                elif selected_values:
                    # A standard select element only allows one option to be
                    # selected, but browsers pick last if somehow multiple.
                    data.append((name, selected_values[-1]))
                elif options:
                    # Pick the first option if none are selected
                    data.append((name, options[0].get("value", "")))
```